Notes/Algorithms/Data Structures/Hash-Tables.c

```diff
@@ -1,230 +1,244 @@
+bool isCollision(HashTable * table, int index) {
+    if (table->items[index] != NULL) {
+        return true;
     }
+    return false;
 }
 
-//compare checks the Items
-bool compare() {
-    for (int i = 0; i < SIZE; i++) {
-        int testKey = testHash[i]->key;
-        int testValue = testHash[i]->value;
-        int expectedKey = expectedHash[i]->key;
-        int expectedValue = expectedHash[i]->value;
+// compare checks the testHash and expectedHash for parity
+// if the two HashTables match returns true
+// if a mismatch occurs returns false
+bool compare(HashTable * testHash, HashTable * expectedHash) {
+    for (int i = 0; i < 10; i++) {
+        int testKey = testHash->items[i]->key;
+        int testValue = testHash->items[i]->value;
+        int expectedKey = expectedHash->items[i]->key;
+        int expectedValue = expectedHash->items[i]->value;
         if (testKey != expectedKey || testValue != expectedValue) {
             return false;
         }
     }
     return true;
 }
 
-int hashing(int key) {
-    return key % SIZE;
+// hashing 
+// returns a "hashed" index based on modulus of the key's ASCII integer sum and the SIZE of the hashTable
+int hashing(HashTable *hashtable, char *key ) {
+    int i = 0;
+    int count = 0;
+    while (key[i] != NULL) {
+        count += key[i];
+        i++;
+    }
+    return count % hashtable->size;
 }
+
+
 // creates a new Item struct
 // returns new struct
-Item * createItem(int key, int value, int success) {
+Item * createItem(char key[], int value, int * success) {
     Item * item = malloc(sizeof(Item));
-    if (item == NULL) {
-    } else {
-        item->key = key;
+    if (item != NULL) {        
+        item->key = strdup(key);
         item->value = value;
-        success = SUCCESS;
+        *success = SUCCESS;
         return item;
     }
+    return NULL;
 }
 
-// insert takes in a key and value
-// calls a helper function to create an Item struct with the provided key and value
-// adds the new Item to the hashTable
-// if the expected value is in the hashTable at the expected key then returns a success code
-void insert(int type, int key, int value, int success) {
-    if (type == TEST) {
-        testHash[key] = createItem(key, value, success);
-        if (testHash[key]->value != value) {
-            success = ERROR;
-        }
-        
-    } else if (type == EXPECTED) {
-        expectedHash[key] = createItem(key, value, success);
-        if (expectedHash[key]->value != value) {
-            success = ERROR;
-        }
+// creates a new HashTable struct
+// returns new HashTable struct
+HashTable * createHashTable(int size) {
+    if (size < 1 || size > 100) return NULL;
+    HashTable * hashTable = malloc(sizeof(HashTable));
+    hashTable->items = malloc(sizeof(Item) * size);
+    if (hashTable == NULL || hashTable->items == NULL) {
+        return NULL;
     }
+    hashTable->size = size;
+    for (int i = 0; i < size; i++) {
+        hashTable->items[i] = NULL;
```

Comment:
Many of your other functions don't check that `items[i]` may be null before dereferencing.  Is this the intended neutral state?

Reply:
It was intended to catch any malloc fails after receiving feedback from Sasha on a previous pull request. Is there times throughout my code where it would be prudent to use a similar logic? (May be a knowledge gap on my part that is missing this).

Reply:
Added checks to see if items[index] != NULL before proceeding with dereferencing ->key and ->value per meeting conversation!
[commit 77f1405ef39e5ee79b413d4e907eabee5c4d29ba](https://github.com/kcarrel/Training-Notes/pull/12/commits/77f1405ef39e5ee79b413d4e907eabee5c4d29ba)